pyanaconda/display.py

```diff
@@ -0,0 +1,328 @@
+#
+# display.py:  graphical display setup for the Anaconda GUI
+#
+# Copyright (C) 2016
+# Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Author(s):  Martin Kolman <mkolman@redhat.com>
+#
+import os
+import subprocess
+import time
+import pkgutil
+
+from pyanaconda.i18n import _
+
+import logging
+log = logging.getLogger("anaconda")
+stdout_log = logging.getLogger("anaconda.stdout")
+
+from pyanaconda import constants
+from pyanaconda import iutil
+from pyanaconda import vnc
+from pyanaconda.flags import flags
+from pyanaconda import isys
+from pyanaconda import startup_utils
+from pyanaconda.nm import nm_is_connected, nm_is_connecting
+
+from pyanaconda.ui.tui.simpleline import App
+from pyanaconda.ui.tui.spokes.askvnc import AskVNCSpoke
+
+# needed for checking if the pyanaconda.ui.gui modules are available
+import pyanaconda.ui
+
+import blivet
+
+from pykickstart.constants import DISPLAY_MODE_TEXT
+
+# Spice
+
+def start_spice_vd_agent():
+    """Start the spice vdagent.
+
+    For certain features to work spice requires that the guest os
+    is running the spice vdagent.
+    """
+    status = iutil.execWithRedirect("spice-vdagent", [])
+    if status:
+        log.info("spice-vdagent exited with status %d", status)
+    else:
+        log.info("Started spice-vdagent.")
+
+# VNC
+
+def ask_vnc_question(anaconda, vnc_server, message):
+    """ Ask the user if TUI or GUI-over-VNC should be started.
+
+    :param anaconda: instance of the Anaconda class
+    :param vnc_server: instance of the VNC server object
+    :param str message: a message to show to the user together
+                        with the question
+    """
+    app = App("VNC Question")
+    spoke = AskVNCSpoke(app, anaconda.ksdata, message)
+    app.schedule_screen(spoke)
+    app.run()
+
+    if anaconda.ksdata.vnc.enabled:
+        if not anaconda.gui_mode:
+            log.info("VNC requested via VNC question, switching Anaconda to GUI mode.")
+        anaconda.display_mode = constants.DISPLAY_MODE_GUI
+        flags.usevnc = True
+        vnc_server.password = anaconda.ksdata.vnc.password
+
+def check_vnc_can_be_started(anaconda):
+    """Check if we can start VNC in the current environment.
+
+    :returns: if VNC can be started and list of possible reasons
+              why VNC can't be started
+    :rtype: bool and list
```

Comment:
I would recommend to change the rtype to ``(boot, list)`` or ``tuple of (bool, list)``.